### ISSUE TITLE
Weird Behavior of reduce

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#308]
Code: https://github.com/ankurankan/pgmpy/tree/weird_reduce

Tests behaving weirdly and fails sometimes and pass sometimes. Something to do with the changes in the `reduce` method.

**Comment by palashahuja** [comment]
@ankurankan  :laughing: Is it that serious ?

**Comment by ankurankan** [comment]
I think we have changed some behavior of reduce because of which map
inference is failing. Will have to look why this is happening. There's
nothing wrong with your changes.

**Comment by palashahuja** [comment]
@ankurankan Also there is the following bug (if reduce is implemented ) ---

```
In [5]:phi = Factor(['x1','x2','x3'],[2,3,2],np.arange(12))

In [6]:phi.reduce(['x1_0','x2_1','x3_1'])
In [7]:phi._index_for_assignment([0,1,1])
Out[7]:    array([2])
```

**Comment by ankurankan** [comment]
Fixed in #316
